File: packages/swift-scanner/Sources/SpecLinkSwiftScanner/Scanner.swift
import Foundation
import SwiftParser
import SwiftSyntax

public final class Scanner {
  public init() {}

  public func scan(requestData: Data) throws -> Data {
    let request = try JSONDecoder().decode(WorkerRequest.self, from: requestData)
    let files = request.files.map { scanFile($0, visibility: request.options.visibility) }
    let response = WorkerResponse(
      schemaVersion: 1,
      requestId: request.requestId,
      language: "swift",
      files: files
    )
    let encoder = JSONEncoder()
    encoder.outputFormatting = [.sortedKeys]
    return try encoder.encode(response)
  }

  private func scanFile(_ file: WorkerFile, visibility: [String]?) -> WorkerFileResponse {
    let tree = Parser.parse(source: file.content)
    if tree.hasError {
      return WorkerFileResponse(
        filePath: file.filePath,
        symbols: [],
        undocumentedSymbols: [],
        links: [],
        diagnostics: [
          diagnostic(
            code: "code_parse_error",
            target: file.filePath,
            message: "Swift parser reported syntax errors.",
            filePath: file.filePath,
            line: 1,
            column: 1
          )
        ]
      )
    }

    let visibilitySet = Set(visibility ?? ["public", "open"])
    let source = SourceText(filePath: file.filePath, content: file.content)
    let declarations = collectDeclarations(source: source, visibility: visibilitySet)

    var symbols: [CodeSymbol] = []
    var undocumentedSymbols: [CodeSymbol] = []
    var links: [DocLink] = []
    var diagnostics: [Diagnostic] = []
    var seenEndpoints = Set<String>()
    var duplicateEndpoints = Set<String>()

    for declaration in declarations {
      if declaration.unsupported {
        diagnostics.append(
          diagnostic(
            code: "unsupported_declaration",
            target: file.filePath,
            message: "Swift declaration annotated with @doc is not supported.",
            filePath: file.filePath,
            line: declaration.line,
            column: declaration.column,
            range: declaration.nameRange
          )
        )
        continue
      }

      guard declaration.visible else { continue }

      let symbol = makeSymbol(filePath: file.filePath, declaration: declaration)
      if declaration.docTargets.isEmpty {
        if !undocumentedSymbols.contains(where: { $0.endpoint == symbol.endpoint }) {
          undocumentedSymbols.append(symbol)
        }
        continue
      }

      if seenEndpoints.contains(symbol.endpoint) {
        if !duplicateEndpoints.contains(symbol.endpoint) {
          diagnostics.append(
            diagnostic(
              code: "duplicate_code_symbol",
              target: symbol.endpoint,
              message: "Duplicate Swift code symbol endpoint: \(symbol.endpoint)",
              filePath: file.filePath,
              line: declaration.line,
              column: declaration.column,
              range: declaration.nameRange
            )
          )
          duplicateEndpoints.insert(symbol.endpoint)
        }
        continue
      }
      seenEndpoints.insert(symbol.endpoint)
      symbols.append(symbol)

      var seenTargets = Set<String>()
      for docTarget in declaration.docTargets where !seenTargets.contains(docTarget.target) {
        seenTargets.insert(docTarget.target)
        links.append(
          DocLink(
            direction: "code-to-doc",
            source: symbol.endpoint,
            target: docTarget.target,
            location: SourceLocation(
              filePath: file.filePath,
              line: docTarget.line,
              column: docTarget.column
            ),
            targetRange: docTarget.range
          )
        )
      }
    }

    return WorkerFileResponse(
      filePath: file.filePath,
      symbols: symbols,
      undocumentedSymbols: undocumentedSymbols,
      links: links,
      diagnostics: diagnostics
    )
  }
}

private struct Declaration {
  let symbolName: String
  let canonicalId: String
  let line: Int
  let column: Int
  let visible: Bool
  let unsupported: Bool
  let nameRange: SourceRange?
  let declarationRange: SourceRange?
  let signatureRange: SourceRange?
  let docTargets: [DocTarget]
}

private struct DocTarget {
  let target: String
  let line: Int
  let column: Int
  let range: SourceRange?
}

private struct TypeScope {
  let name: String
  let closesAtDepth: Int
}

private func collectDeclarations(source: SourceText, visibility: Set<String>) -> [Declaration] {
  var declarations: [Declaration] = []
  var pendingDoc: PendingDoc?
  var scopes: [TypeScope] = []
  var braceDepth = 0
  var inBlockDoc = false
  var blockDocLines: [String] = []
  var blockDocStart = 1

  for lineNumber in 1...source.lineCount {
    let line = source.line(lineNumber)
    let trimmed = line.trimmingCharacters(in: .whitespaces)

    while let last = scopes.last, braceDepth < last.closesAtDepth {
      scopes.removeLast()
    }

    if inBlockDoc {
      blockDocLines.append(line)
      if trimmed.contains("*/") {
        pendingDoc = parseDocTargets(
          lines: blockDocLines,
          startLine: blockDocStart,
          source: source
        )
        blockDocLines = []
        inBlockDoc = false
      }
      braceDepth += braceDelta(line)
      continue
    }

    if trimmed.hasPrefix("/**") {
      inBlockDoc = !trimmed.contains("*/")
      blockDocStart = lineNumber
      blockDocLines = [line]
      if !inBlockDoc {
        pendingDoc = parseDocTargets(lines: blockDocLines, startLine: blockDocStart, source: source)
        blockDocLines = []
      }
      braceDepth += braceDelta(line)
      continue
    }

    if trimmed.hasPrefix("///") {
      let current = pendingDoc ?? PendingDoc(targets: [], startLine: lineNumber)
      let targets = current.targets + parseDocLine(line, lineNumber: lineNumber, source: source)
      pendingDoc = PendingDoc(targets: targets, startLine: current.startLine)
      braceDepth += braceDelta(line)
      continue
    }

    if trimmed.isEmpty || trimmed.hasPrefix("//") {
      braceDepth += braceDelta(line)
      continue
    }

    if let info = parseSupportedDeclaration(line: line, typeName: scopes.last?.name) {
      let doc = pendingDoc
      pendingDoc = nil
      let visible = visibility.contains(info.visibility)
      let declarationRange = source.range(
        startLine: doc?.startLine ?? lineNumber,
        startColumn: 1,
        endLine: lineNumber,
        endColumn: source.utf16ColumnAfterLine(lineNumber)
      )
      declarations.append(
        Declaration(
          symbolName: info.symbolName,
          canonicalId: info.canonicalId,
          line: lineNumber,
          column: info.nameColumn,
          visible: visible,
          unsupported: false,
          nameRange: source.range(
            startLine: lineNumber,
            startColumn: info.nameColumn,
            endLine: lineNumber,
            endColumn: info.nameColumn + info.symbolName.utf16.count
          ),
          declarationRange: declarationRange,
          signatureRange: declarationRange,
          docTargets: doc?.targets ?? []
        )
      )
      if let typeName = info.opensTypeScope, line.contains("{") {
        scopes.append(TypeScope(name: typeName, closesAtDepth: braceDepth + 1))
      }
      braceDepth += braceDelta(line)
      continue
    }

    if let extended = parseExtension(line: line), line.contains("{") {
      pendingDoc = nil
      scopes.append(TypeScope(name: extended, closesAtDepth: braceDepth + 1))
      braceDepth += braceDelta(line)
      continue
    }

    if let doc = pendingDoc {
      declarations.append(
        Declaration(
          symbolName: "",
          canonicalId: "",
          line: lineNumber,
          column: firstNonWhitespaceColumn(line),
          visible: true,
          unsupported: true,
          nameRange: nil,
          declarationRange: nil,
          signatureRange: nil,
          docTargets: doc.targets
        )
      )
      pendingDoc = nil
    }

    braceDepth += braceDelta(line)
  }

  return declarations
}

private struct DeclarationInfo {
  let symbolName: String
  let canonicalId: String
  let visibility: String
  let nameColumn: Int
  let opensTypeScope: String?
}

private func parseSupportedDeclaration(line: String, typeName: String?) -> DeclarationInfo? {
  let pattern = #"^\s*(?:(open|public|internal|fileprivate|private)\s+)?(?:(?:final|static|class|mutating|nonmutating|override|required|convenience)\s+)*(class|struct|enum|protocol|actor|func|var|let|init)\b\s*([A-Za-z_][A-Za-z0-9_]*)?"#
  guard let match = line.firstMatch(pattern: pattern) else { return nil }
  let visibility = match.group(1) ?? "internal"
  let kind = match.group(2) ?? ""
  let rawName = match.group(3)

  if ["class", "struct", "enum", "protocol", "actor"].contains(kind), let name = rawName {
    return DeclarationInfo(
      symbolName: name,
      canonicalId: typeName.map { "\($0).\(name)" } ?? name,
      visibility: visibility,
      nameColumn: utf16Column(of: name, in: line),
      opensTypeScope: name
    )
  }

  if kind == "init" {
    let labels = argumentLabels(line: line)
    let canonical = "\(typeName.map { "\($0)." } ?? "")init(\(labels))"
    return DeclarationInfo(
      symbolName: "init",
      canonicalId: canonical,
      visibility: visibility,
      nameColumn: utf16Column(of: "init", in: line),
      opensTypeScope: nil
    )
  }

  guard let name = rawName else { return nil }
  let memberPrefix = typeName.map { "\($0)." } ?? ""
  let suffix = kind == "func" ? "(\(argumentLabels(line: line)))" : ""
  return DeclarationInfo(
    symbolName: name,
    canonicalId: "\(memberPrefix)\(name)\(suffix)",
    visibility: visibility,
    nameColumn: utf16Column(of: name, in: line),
    opensTypeScope: nil
  )
}

private func parseExtension(line: String) -> String? {
  let pattern = #"^\s*(?:(?:public|internal|fileprivate|private)\s+)?extension\s+([A-Za-z_][A-Za-z0-9_\.]*)"#
  return line.firstMatch(pattern: pattern)?.group(1)
}

private func argumentLabels(line: String) -> String {
  guard let open = line.firstIndex(of: "("), let close = line[open...].firstIndex(of: ")") else {
    return ""
  }
  let params = line[line.index(after: open)..<close]
  if params.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
    return ""
  }
  return params.split(separator: ",").map { raw in
    let parameter = raw.trimmingCharacters(in: .whitespacesAndNewlines)
    let parts = parameter.split(whereSeparator: { $0 == " " || $0 == "\t" })
    if parts.first == "_" {
      return "_:"
    }
    let label = (parts.first ?? "").trimmingCharacters(in: CharacterSet(charactersIn: ":"))
    return "\(label):"
  }.joined()
}

private struct PendingDoc {
  let targets: [DocTarget]
  let startLine: Int
}

private func parseDocTargets(lines: [String], startLine: Int, source: SourceText) -> PendingDoc {
  var targets: [DocTarget] = []
  for (offset, line) in lines.enumerated() {
    targets.append(contentsOf: parseDocLine(line, lineNumber: startLine + offset, source: source))
  }
  return PendingDoc(targets: targets, startLine: startLine)
}

private func parseDocLine(_ line: String, lineNumber: Int, source: SourceText) -> [DocTarget] {
  guard let range = line.range(of: #"@doc\s+(\S+)"#, options: .regularExpression) else {
    return []
  }
  let matched = String(line[range])
  let target = matched.replacingOccurrences(
    of: #"^@doc\s+"#,
    with: "",
    options: .regularExpression
  )
  let start = line.distance(from: line.startIndex, to: range.lowerBound) + matched.utf16.count - target.utf16.count + 1
  return [
    DocTarget(
      target: target,
      line: lineNumber,
      column: start,
      range: source.range(
        startLine: lineNumber,
        startColumn: start,
        endLine: lineNumber,
        endColumn: start + target.utf16.count
      )
    )
  ]
}

private func makeSymbol(filePath: String, declaration: Declaration) -> CodeSymbol {
  CodeSymbol(
    kind: "code",
    language: "swift",
    filePath: filePath,
    symbolName: declaration.symbolName,
    canonicalId: declaration.canonicalId,
    endpoint: "\(filePath)#\(declaration.canonicalId)",
    location: SourceLocation(filePath: filePath, line: declaration.line, column: declaration.column),
    nameRange: declaration.nameRange,
    declarationRange: declaration.declarationRange,
    signatureRange: declaration.signatureRange
  )
}

private func diagnostic(
  code: String,
  target: String,
  message: String,
  filePath: String,
  line: Int,
  column: Int,
  range: SourceRange? = nil
) -> Diagnostic {
  Diagnostic(
    severity: "error",
    code: code,
    target: target,
    language: "swift",
    source: nil,
    message: message,
    location: SourceLocation(filePath: filePath, line: line, column: column),
    range: range
  )
}

private func braceDelta(_ line: String) -> Int {
  var delta = 0
  var inString = false
  for char in line {
    if char == "\"" { inString.toggle() }
    if inString { continue }
    if char == "{" { delta += 1 }
    if char == "}" { delta -= 1 }
  }
  return delta
}

private func firstNonWhitespaceColumn(_ line: String) -> Int {
  var column = 1
  for char in line {
    if char != " " && char != "\t" { return column }
    column += String(char).utf16.count
  }
  return 1
}

private func utf16Column(of needle: String, in line: String) -> Int {
  guard let range = line.range(of: needle) else { return firstNonWhitespaceColumn(line) }
  return line[..<range.lowerBound].utf16.count + 1
}

private struct RegexMatch {
  let match: NSTextCheckingResult
  let text: NSString

  func group(_ index: Int) -> String? {
    let range = match.range(at: index)
    guard range.location != NSNotFound else { return nil }
    return text.substring(with: range)
  }
}

private extension String {
  func firstMatch(pattern: String) -> RegexMatch? {
    guard let regex = try? NSRegularExpression(pattern: pattern) else { return nil }
    let text = self as NSString
    let range = NSRange(location: 0, length: text.length)
    guard let match = regex.firstMatch(in: self, range: range) else { return nil }
    return RegexMatch(match: match, text: text)
  }
}

private struct SourceText {
  let filePath: String
  let lines: [String]

  init(filePath: String, content: String) {
    self.filePath = filePath
    self.lines = content.split(separator: "\n", omittingEmptySubsequences: false).map(String.init)
  }

  var lineCount: Int { lines.count }

  func line(_ number: Int) -> String {
    guard number > 0 && number <= lines.count else { return "" }
    return lines[number - 1]
  }

  func utf16ColumnAfterLine(_ number: Int) -> Int {
    line(number).utf16.count + 1
  }

  func range(startLine: Int, startColumn: Int, endLine: Int, endColumn: Int) -> SourceRange {
    SourceRange(
      start: Position(line: startLine, column: startColumn),
      end: Position(line: endLine, column: endColumn)
    )
  }
}

File: packages/dart-scanner/lib/scanner.dart
import 'dart:convert';

import 'package:analyzer/dart/analysis/utilities.dart';
import 'package:analyzer/dart/ast/ast.dart';
import 'package:analyzer/dart/ast/token.dart';
import 'package:analyzer/source/line_info.dart';

import 'models.dart';

/// Scans Dart source files for SpecLink `@doc` annotations using the Dart
/// `analyzer` AST.
///
/// Dart has no method overloading, so member canonical IDs are type-qualified
/// names without parameter signatures (`AuthService.login`). Setters carry a
/// trailing `=` to stay distinct from a same-named getter or field, and the
/// unnamed constructor is `Type.new`.
class Scanner {
  String scan(String requestJson) {
    final request =
        WorkerRequest.fromJson(jsonDecode(requestJson) as Map<String, dynamic>);
    final files = request.files
        .map((file) => _scanFile(file, request.options.visibility))
        .toList();
    final response = WorkerResponse(
      schemaVersion: 1,
      requestId: request.requestId,
      language: 'dart',
      files: files,
    );
    return jsonEncode(response.toJson());
  }

  WorkerFileResponse _scanFile(WorkerFile file, List<String>? visibility) {
    final parsed = parseString(content: file.content, throwIfDiagnostics: false);
    if (parsed.errors.isNotEmpty) {
      return WorkerFileResponse(
        filePath: file.filePath,
        symbols: const [],
        undocumentedSymbols: const [],
        links: const [],
        diagnostics: [
          Diagnostic(
            code: 'code_parse_error',
            target: file.filePath,
            message: 'Dart analyzer reported syntax errors.',
            location: SourceLocation(filePath: file.filePath, line: 1, column: 1),
          ),
        ],
      );
    }

    final collector = _Collector(file.content, parsed.lineInfo);
    collector.visitUnit(parsed.unit);

    return _build(file.filePath, collector.declarations);
  }

  WorkerFileResponse _build(String filePath, List<_Declaration> declarations) {
    final symbols = <CodeSymbol>[];
    final undocumented = <CodeSymbol>[];
    final links = <DocLink>[];
    final diagnostics = <Diagnostic>[];
    final seenEndpoints = <String>{};
    final duplicateEndpoints = <String>{};

    for (final declaration in declarations) {
      if (declaration.unsupported) {
        diagnostics.add(
          Diagnostic(
            code: 'unsupported_declaration',
            target: filePath,
            message: 'Dart declaration annotated with @doc is not supported.',
            location: SourceLocation(
              filePath: filePath,
              line: declaration.line,
              column: declaration.column,
            ),
            range: declaration.nameRange,
          ),
        );
        continue;
      }

      if (!declaration.visible) continue;

      final symbol = declaration.toSymbol(filePath);
      if (declaration.docTargets.isEmpty) {
        if (!undocumented.any((s) => s.endpoint == symbol.endpoint)) {
          undocumented.add(symbol);
        }
        continue;
      }

      if (seenEndpoints.contains(symbol.endpoint)) {
        if (!duplicateEndpoints.contains(symbol.endpoint)) {
          diagnostics.add(
            Diagnostic(
              code: 'duplicate_code_symbol',
              target: symbol.endpoint,
              message: 'Duplicate Dart code symbol endpoint: ${symbol.endpoint}',
              location: SourceLocation(
                filePath: filePath,
                line: declaration.line,
                column: declaration.column,
              ),
              range: declaration.nameRange,
            ),
          );
          duplicateEndpoints.add(symbol.endpoint);
        }
        continue;
      }
      seenEndpoints.add(symbol.endpoint);
      symbols.add(symbol);

      final seenTargets = <String>{};
      for (final docTarget in declaration.docTargets) {
        if (seenTargets.contains(docTarget.target)) continue;
        seenTargets.add(docTarget.target);
        links.add(
          DocLink(
            source: symbol.endpoint,
            target: docTarget.target,
            location: SourceLocation(
              filePath: filePath,
              line: docTarget.line,
              column: docTarget.column,
            ),
            targetRange: docTarget.range,
          ),
        );
      }
    }

    return WorkerFileResponse(
      filePath: filePath,
      symbols: symbols,
      undocumentedSymbols: undocumented,
      links: links,
      diagnostics: diagnostics,
    );
  }
}

class _Declaration {
  _Declaration({
    required this.symbolName,
    required this.canonicalId,
    required this.line,
    required this.column,
    required this.visible,
    required this.unsupported,
    required this.docTargets,
    this.nameRange,
    this.declarationRange,
    this.signatureRange,
  });

  final String symbolName;
  final String canonicalId;
  final int line;
  final int column;
  final bool visible;
  final bool unsupported;
  final List<_DocTarget> docTargets;
  final SourceRange? nameRange;
  final SourceRange? declarationRange;
  final SourceRange? signatureRange;

  CodeSymbol toSymbol(String filePath) => CodeSymbol(
        filePath: filePath,
        symbolName: symbolName,
        canonicalId: canonicalId,
        endpoint: '$filePath#$canonicalId',
        location: SourceLocation(filePath: filePath, line: line, column: column),
        nameRange: nameRange,
        declarationRange: declarationRange,
        signatureRange: signatureRange,
      );
}

class _DocTarget {
  _DocTarget({
    required this.target,
    required this.line,
    required this.column,
    required this.range,
  });

  final String target;
  final int line;
  final int column;
  final SourceRange range;
}

class _Collector {
  _Collector(this.content, this.lineInfo);

  final String content;
  final LineInfo lineInfo;
  final List<_Declaration> declarations = [];

  void visitUnit(CompilationUnit unit) {
    for (final declaration in unit.declarations) {
      _handleTopLevel(declaration);
    }
  }

  void _handleTopLevel(CompilationUnitMember declaration) {
    if (declaration is FunctionDeclaration) {
      final name = declaration.name.lexeme;
      _record(
        nameToken: declaration.name,
        node: declaration,
        symbolName: name,
        canonicalId: declaration.isSetter ? '$name=' : name,
        signatureEnd: declaration.functionExpression.body.offset,
      );
      return;
    }

    if (declaration is TopLevelVariableDeclaration) {
      for (final variable in declaration.variables.variables) {
        _record(
          nameToken: variable.name,
          node: declaration,
          symbolName: variable.name.lexeme,
          canonicalId: variable.name.lexeme,
          signatureEnd: declaration.end,
        );
      }
      return;
    }

    if (declaration is ClassDeclaration) {
      _recordType(declaration.name, declaration, declaration.leftBracket);
      _handleMembers(declaration.members, declaration.name.lexeme);
      return;
    }

    if (declaration is EnumDeclaration) {
      _recordType(declaration.name, declaration, declaration.leftBracket);
      _handleMembers(declaration.members, declaration.name.lexeme);
      return;
    }

    if (declaration is MixinDeclaration) {
      _recordType(declaration.name, declaration, declaration.leftBracket);
      _handleMembers(declaration.members, declaration.name.lexeme);
      return;
    }

    if (declaration is ExtensionDeclaration) {
      // Extensions produce no symbol; their members canonicalize against the
      // extended type. A `@doc` on the extension itself is ignored.
      final qualifier = _extendedTypeName(declaration);
      if (qualifier.isNotEmpty) {
        _handleMembers(declaration.members, qualifier);
      }
      return;
    }

    _markUnsupportedIfAnnotated(declaration);
  }

  void _handleMembers(List<ClassMember> members, String qualifier) {
    for (final member in members) {
      if (member is MethodDeclaration) {
        if (member.isOperator) continue;
        final name = member.name.lexeme;
        _record(
          nameToken: member.name,
          node: member,
          symbolName: name,
          canonicalId: member.isSetter ? '$qualifier.$name=' : '$qualifier.$name',
          signatureEnd: member.body.offset,
        );
      } else if (member is FieldDeclaration) {
        for (final variable in member.fields.variables) {
          _record(
            nameToken: variable.name,
            node: member,
            symbolName: variable.name.lexeme,
            canonicalId: '$qualifier.${variable.name.lexeme}',
            signatureEnd: member.end,
          );
        }
      } else if (member is ConstructorDeclaration) {
        final ctorName = member.name?.lexeme;
        final nameToken = member.name ?? member.returnType.beginToken;
        _record(
          nameToken: nameToken,
          node: member,
          symbolName: ctorName ?? 'new',
          canonicalId:
              ctorName != null ? '$qualifier.$ctorName' : '$qualifier.new',
          signatureEnd: member.body.offset,
        );
      }
    }
  }

  void _recordType(Token nameToken, AnnotatedNode node, Token leftBracket) {
    _record(
      nameToken: nameToken,
      node: node,
      symbolName: nameToken.lexeme,
      canonicalId: nameToken.lexeme,
      signatureEnd: leftBracket.offset,
    );
  }

  void _record({
    required Token nameToken,
    required AnnotatedNode node,
    required String symbolName,
    required String canonicalId,
    required int signatureEnd,
  }) {
    final visible = !symbolName.startsWith('_');
    final declStart = _declarationStart(node);
    final location = lineInfo.getLocation(nameToken.offset);
    declarations.add(
      _Declaration(
        symbolName: symbolName,
        canonicalId: canonicalId,
        line: location.lineNumber,
        column: location.columnNumber,
        visible: visible,
        unsupported: false,
        nameRange: _range(nameToken.offset, nameToken.end),
        declarationRange: _range(declStart, node.end),
        signatureRange: _range(declStart, signatureEnd),
        docTargets: _docTargets(node),
      ),
    );
  }

  void _markUnsupportedIfAnnotated(AnnotatedNode node) {
    final targets = _docTargets(node);
    if (targets.isEmpty) return;
    final start = node.firstTokenAfterCommentAndMetadata.offset;
    final location = lineInfo.getLocation(start);
    declarations.add(
      _Declaration(
        symbolName: '',
        canonicalId: '',
        line: location.lineNumber,
        column: location.columnNumber,
        visible: true,
        unsupported: true,
        docTargets: targets,
      ),
    );
  }

  String _extendedTypeName(ExtensionDeclaration extension) {
    final onType = extension.onClause?.extendedType;
    if (onType == null) return '';
    return onType.toSource().split('<').first.trim();
  }

  int _declarationStart(AnnotatedNode node) {
    final comment = node.documentationComment;
    if (comment != null) return comment.offset;
    return node.firstTokenAfterCommentAndMetadata.offset;
  }

  List<_DocTarget> _docTargets(AnnotatedNode node) {
    final comment = node.documentationComment;
    if (comment == null) return const [];
    final text = content.substring(comment.offset, comment.end);
    final targets = <_DocTarget>[];
    for (final match in RegExp(r'@doc\s+(\S+)').allMatches(text)) {
      final target = match.group(1)!;
      final targetStartInText = match.start + (match.group(0)!.length - target.length);
      final location = lineInfo.getLocation(comment.offset + targetStartInText);
      targets.add(
        _DocTarget(
          target: target,
          line: location.lineNumber,
          column: location.columnNumber,
          range: SourceRange(
            start: Position(line: location.lineNumber, column: location.columnNumber),
            end: Position(
              line: location.lineNumber,
              column: location.columnNumber + target.length,
            ),
          ),
        ),
      );
    }
    return targets;
  }

  Position _position(int offset) {
    final location = lineInfo.getLocation(offset);
    return Position(line: location.lineNumber, column: location.columnNumber);
  }

  SourceRange _range(int start, int end) =>
      SourceRange(start: _position(start), end: _position(end));
}
